### ISSUE TITLE
Single device login

### ISSUE THREAD

**Issue by jslim89** [tymondesigns/jwt-auth#1531]
### Subject of the issue
I want to allow 1 user can only login via 1 device.
e.g.
User 1 login from device 1
User 1 login from device 2 (system will force the device 1 to logout)

### Your environment
| Q                 | A
| ----------------- | ---
| Bug?              | no 
| New Feature?      | yes
| Framework         | Laravel
| Framework version | 5.5
| Package version   | dev-develop
| PHP version       | 7.2

**Comment by ngocnd0242** [comment]
How to resolved follow bellow:
1. Store token when use logined
2. Invalidate token on next time login
```
 JWTAuth::setToken($user->lastest_token);
 JWTAuth::invalidate();
```

**Comment by TroyRay** [comment]
Thank you! It works for me. @ngodinhngoc

**Comment by roman-briazgalov** [comment]
@ngocnd0242, @TroyRay
Hi!
I have the same problem.
Could you please tell what version of JWT-auth you use? And where exactly should i add this code to make users be logged in only from one device?

**Comment by itsensepk** [comment]
@TroyRay @ngocnd0242 @jslim89 @roman-briazgalov @MitchellMcKenna 
Hi, 
How & where do you implement this ? I couldnt figure it out in Laravel 5.7, 

I have basic login controller, tried implementing there but failed. help needed

**Comment by Amelbo3** [comment]
Hello there,
I am new here and i have a question. Is is possible to get token from 2
different variables other than email and password? Please help!

**Comment by roman-briazgalov** [comment]
> @TroyRay @ngocnd0242 @jslim89 @roman-briazgalov @MitchellMcKenna
> Hi,
> How & where do you implement this ? I couldnt figure it out in Laravel 5.7,
> 
> I have basic login controller, tried implementing there but failed. help needed

Hi!
i made it by Laravel methods:
https://laravel.com/docs/5.8/authentication#invalidating-sessions-on-other-devices

I have just added this code in LoginController:
`public function authenticated(Request $request)`
`{`
`        Auth::logoutOtherDevices($request->password);`
`}`

**Comment by itsensepk** [comment]
@roman-briazgalov hey, can you share your auth controller, I dont know why i'm failing to do that.. 
may be my brain is exhausted by continuous work,
Really appreciate it!

**Comment by Rodolfo-XDR** [comment]
@roman-briazgalov  I don't get how you managed to use normal LoginController methods implementing JWT. The authenticated function is only triggered if you don't override your login function, or if you return "sendLoginresponse()"

`return $this->sendLoginResponse($request);`

**Comment by ngocnd0242** [comment]
> @ngocnd0242, @TroyRay
> Hi!
> I have the same problem.
> Could you please tell what version of JWT-auth you use? And where exactly should i add this code to make users be logged in only from one device?

Sorry for inlate reply, I used this version `"tymon/jwt-auth": "1.0.0-rc.2"`

**Comment by ngocnd0242** [comment]
> > @TroyRay @ngocnd0242 @jslim89 @roman-briazgalov @MitchellMcKenna
> > Hi,
> > How & where do you implement this ? I couldnt figure it out in Laravel 5.7,
> > I have basic login controller, tried implementing there but failed. help needed
> 
> Hi!
> i made it by Laravel methods:
> https://laravel.com/docs/5.8/authentication#invalidating-sessions-on-other-devices
> 
> I have just added this code in LoginController:
> `public function authenticated(Request $request)`
> `{`
> ` Auth::logoutOtherDevices($request->password);`
> `}`

@roman-briazgalov sorry for inlate reply, I used for `"laravel/framework": "5.6.*",`

**Comment by francsiswanto** [comment]
Hi.

This is my hack to solve SSO, I take below script from https://jwt-auth.readthedocs.io/en/develop/quick-start/ as starting point, and I'm using this module with version 1.02 (Nov 2020), also make sure blacklist feature is enabled.
```
namespace App\Http\Controllers;

use Illuminate\Support\Facades\Auth;
use App\Http\Controllers\Controller;

class AuthController extends Controller
{
    /**
     * Get a JWT via given credentials.
     *
     * @return \Illuminate\Http\JsonResponse
     */
    public function login()
    {
        $credentials = request(['email', 'password']);

        if (! $token = auth()->attempt($credentials)) {
            return response()->json(['error' => 'Unauthorized'], 401);
        }
        
+       // Get "sub" value from $token (maybe you can use request()->email or maybe not)
+       $sub = json_decode(base64_decode(explode('.', $token)[1]), true)['sub'];
+       $keycache = 'sso:' . $sub;
+
+       // Blacklist the old token if exist
+       $oldToken = \Cache::get($keycache);
+       if ($oldToken) {
+           \JWTAuth::blacklist()
+               ->add(\JWTAuth::manager()
+               ->decode(new \Tymon\JWTAuth\Token($oldToken), FALSE));
+       }
+       
+       // Save current token to cache
+       Cache::put($keycache, $token);

        return $this->respondWithToken($token);
    }
}
```

It works for me.

**Comment by amrshakya** [comment]
> Hi.
> 
> This is my hack to solve SSO, I take below script from https://jwt-auth.readthedocs.io/en/develop/quick-start/ as starting point, and I'm using this module with version 1.02 (Nov 2020), also make sure blacklist feature is enabled.
> 
> ```
> namespace App\Http\Controllers;
> 
> use Illuminate\Support\Facades\Auth;
> use App\Http\Controllers\Controller;
> 
> class AuthController extends Controller
> {
>     /**
>      * Get a JWT via given credentials.
>      *
>      * @return \Illuminate\Http\JsonResponse
>      */
>     public function login()
>     {
>         $credentials = request(['email', 'password']);
> 
>         if (! $token = auth()->attempt($credentials)) {
>             return response()->json(['error' => 'Unauthorized'], 401);
>         }
>         
> +       // Get "sub" value from $token (maybe you can use request()->email or maybe not)
> +       $sub = json_decode(base64_decode(explode('.', $token)[1]), true)['sub'];
> +       $keycache = 'sso:' . $sub;
> +
> +       // Blacklist the old token if exist
> +       $oldToken = \Cache::get($keycache);
> +       if ($oldToken) {
> +           \JWTAuth::blacklist()
> +               ->add(\JWTAuth::manager()
> +               ->decode(new \Tymon\JWTAuth\Token($oldToken), FALSE));
> +       }
> +       
> +       // Save current token to cache
> +       Cache::put($keycache, $token);
> 
>         return $this->respondWithToken($token);
>     }
> }
> ```
> 
> It works for me.



> Hi.
> 
> This is my hack to solve SSO, I take below script from https://jwt-auth.readthedocs.io/en/develop/quick-start/ as starting point, and I'm using this module with version 1.02 (Nov 2020), also make sure blacklist feature is enabled.
> 
> ```
> namespace App\Http\Controllers;
> 
> use Illuminate\Support\Facades\Auth;
> use App\Http\Controllers\Controller;
> 
> class AuthController extends Controller
> {
>     /**
>      * Get a JWT via given credentials.
>      *
>      * @return \Illuminate\Http\JsonResponse
>      */
>     public function login()
>     {
>         $credentials = request(['email', 'password']);
> 
>         if (! $token = auth()->attempt($credentials)) {
>             return response()->json(['error' => 'Unauthorized'], 401);
>         }
>         
> +       // Get "sub" value from $token (maybe you can use request()->email or maybe not)
> +       $sub = json_decode(base64_decode(explode('.', $token)[1]), true)['sub'];
> +       $keycache = 'sso:' . $sub;
> +
> +       // Blacklist the old token if exist
> +       $oldToken = \Cache::get($keycache);
> +       if ($oldToken) {
> +           \JWTAuth::blacklist()
> +               ->add(\JWTAuth::manager()
> +               ->decode(new \Tymon\JWTAuth\Token($oldToken), FALSE));
> +       }
> +       
> +       // Save current token to cache
> +       Cache::put($keycache, $token);
> 
>         return $this->respondWithToken($token);
>     }
> }
> ```
> 
> It works for me.

@francsiswanto  Won't this increase the memory consumption for the cache driver? and which cache driver you are using is it redis?

**Comment by francsiswanto** [comment]
@amrshakya , yes I'm using redis cache, and consumption of cache storage will increase according to the number of users logged in multiply by token length (approximately).

**Comment by amrshakya** [comment]
> @amrshakya , yes I'm using redis cache, and consumption of cache storage will increase according to the number of users logged in multiply by token length (approximately).

So is there anyway you handle for consumption for cache? How do you flush old tokens?

**Comment by francsiswanto** [comment]
> > @amrshakya , yes I'm using redis cache, and consumption of cache storage will increase according to the number of users logged in multiply by token length (approximately).
> 
> So is there anyway you handle for consumption for cache? How do you flush old tokens?

Since I’m using redis, blacklist() function will save blacklist payload with TTL into redis.

**Comment by amrshakya** [comment]
> > > @amrshakya , yes I'm using redis cache, and consumption of cache storage will increase according to the number of users logged in multiply by token length (approximately).
> > 
> > 
> > So is there anyway you handle for consumption for cache? How do you flush old tokens?
> 
> Since I’m using redis, blacklist() function will save blacklist payload with TTL into redis.

Could you provide some code snippets for reference

**Comment by amrshakya** [comment]
> > > @amrshakya , yes I'm using redis cache, and consumption of cache storage will increase according to the number of users logged in multiply by token length (approximately).
> > 
> > 
> > So is there anyway you handle for consumption for cache? How do you flush old tokens?
> 
> Since I’m using redis, blacklist() function will save blacklist payload with TTL into redis.

Thanks
